fud2/fud-core/src/exec/plan/egg_planner.rs

```diff
@@ -0,0 +1,229 @@
+    /// represents the absense of all states and all ops. A term `(root (states 3 x x) (ops x 2))`
+    /// represents a a pair of sets, the first containing a single state with `StateRef` 3 and the
+    /// second containing a single op with `OpRef` 2.
+    enum StateLanguage {
+        // The root of a term. This is used to store a pair of "states" and "ops".
+        "root" = Root([Id; 2]),
+        // A list of states.
+        "states" = States(Box<[Id]>),
+        // A list of ops.
+        "ops" = Ops(Box<[Id]>),
+        // Symbolizes the absense of a state or op.
+        "xxx" = X,
+        // A ref, refering to either a StateRef or OpRef depending on context.
+        Ref(u32),
+    }
+}
+
+/// Construct string which can be parsed to a `root` term in `StateLanguage`.
+///
+/// `states` are in included states in the term, and `through` are the included ops. If
+/// `states_use_x` is set, then all absent states will be marked with `x`. If it is not set, these
+/// unused states will instead be marked with a unique variable to create a string parsable into a
+/// pattern. `ops_use_x` is similar but applies to unused ops.
+///
+/// `all_states` is an ordered list of all states. `all_ops` is an ordered list of all ops.
+fn language_string(
+    states: &[StateRef],
+    through: &[OpRef],
+    states_use_x: bool,
+    ops_use_x: bool,
+    all_states: &[StateRef],
+    all_ops: &[OpRef],
+) -> String {
+    // Collect states into a string.
+    let state_str = all_states
+        .iter()
+        .enumerate()
+        .map(|(i, s)| {
+            if states.contains(s) {
+                s.as_u32().to_string()
+            } else if states_use_x {
+                String::from("xxx")
+            } else {
+                format!("?s{}", i)
+            }
+        })
+        .collect::<Vec<_>>()
+        .join(" ");
+    let state_str = format!("(states {})", state_str);
+
+    // Collect ops into a string.
+    let op_str = all_ops
+        .iter()
+        .enumerate()
+        .map(|(i, o)| {
+            if through.contains(o) {
+                o.as_u32().to_string()
+            } else if ops_use_x {
+                String::from("xxx")
+            } else {
+                format!("?o{}", i)
+            }
+        })
+        .collect::<Vec<_>>()
+        .join(" ");
+    let op_str = format!("(ops {})", op_str);
+
+    // Put them together with root.
+    format!("(root {} {})", state_str, op_str)
+}
+
+/// Creates a `Rewrite` corresponding to `op` and `op_ref` for `StateLanguage`.
+///
+/// `through` is all of the ops required to be used by a plan the planner is attempting to find.
+/// This is used as an optimization so only the ops specified in `through` are needed to be kept
+/// track of in `ops` terms.
+///
+/// `all_states` is an ordered list of all states. `all_ops` is an ordered list of all ops.
+fn rewrite_from_op(
+    op_ref: OpRef,
+    op: &Operation,
```

Comment:
At least for me, I find thinking about a single op is easier even if it means passing a little more mess.

For the second part, forcing planners to have the op graph works fine, and while it makes stuff like this a bit easier, I feel like it's more clear right now to just have them be simple objects wrapping the one function instead of like an object symbolizing a "request" for a plan that has to be evaluated.

Also as planners are part of requests, I don't love the idea of requiring requests to duplicate the op graph from the driver. Of course this could be solved by moving the planner to the driver, but I don't really thing it would cause too much improvement (especially as it makes the `Driver` depend even more on the arguments passed during to the cli).

Reply:
All sounds reasonable! Totally good to ignore this suggestion. :smiley: